src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,1168 @@
+{
+    return data;
+}
+
+FolderMetadata::FolderMetadata(AccountPtr account, const QByteArray& metadata) : _account(account), _metadata(metadata)
+{
+    if (metadata.isEmpty()) {
+        qCInfo(lcCse()) << "Setupping Empty Metadata";
+        setupEmptyMetadata();
+    } else {
+        qCInfo(lcCse()) << "Metadata already exists, deal with it later.";
+    }
+}
+
+// RSA/ECB/OAEPWithSHA-256AndMGF1Padding using private / public key.
+std::string FolderMetadata::encryptMetadataKeys(const nlohmann::json& metadataKeys) const {
+    std::string metadata = metadataKeys.dump();
+    unsigned char *metadataPtr = (unsigned char *) metadata.c_str();
+    size_t metadataPtrLen = metadata.size();
+
+    int err = -1;
+    const int rsaOeapPadding = 1;
+    EVP_PKEY *key = nullptr;
+
+    /*TODO: Verify if we need to setup a RSA engine.
+     * by default it's RSA OEAP */
+    ENGINE *eng = nullptr;
+
+    auto path = publicKeyPath(_account);
+    const char *pathC = qPrintable(path);
+
+    FILE* pkeyFile = fopen(pathC, "r");
+    if (!pkeyFile) {
+        qCInfo(lcCse()) << "Could not open the public key";
+        exit(1);
+    }
+
+    key = PEM_read_PUBKEY(pkeyFile, NULL, NULL, NULL);
+
+    auto ctx = EVP_PKEY_CTX_new(key, eng);
+    if (!ctx) {
+        qCInfo(lcCse()) << "Could not initialize the pkey context.";
+        exit(1);
+    }
+
+    err = EVP_PKEY_encrypt_init(ctx);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Error initilaizing the encryption.";
+        exit(1);
+    }
+
+    err = EVP_PKEY_CTX_set_rsa_padding(ctx, rsaOeapPadding);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Error setting the encryption padding.";
+        exit(1);
+    }
+
+    size_t outLen = 0;
+    err = EVP_PKEY_encrypt(ctx, NULL, &outLen, metadataPtr, metadataPtrLen);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Error retrieving the size of the encrypted data";
+        exit(1);
+    } else {
+        qCInfo(lcCse()) << "Encrption Length:" << outLen;
+    }
+
+    unsigned char *out = (uchar*) OPENSSL_malloc(outLen);
+    if (!out) {
+        qCInfo(lcCse()) << "Error requesting memory for the encrypted contents";
+        exit(1);
+    }
+
+    err = EVP_PKEY_encrypt(ctx, out, &outLen, metadataPtr, metadataPtrLen);
+    if (err <= 0) {
+        qCInfo(lcCse()) << "Could not encrypt key." << err;
+        exit(1);
+    }
+
+    // Transform the encrypted data into base64.
+    const auto raw = QByteArray((const char*) out, outLen);
+    const auto b64 = raw.toBase64();
+    const auto ret = std::string(b64.constData(), b64.length());
+
+    qCInfo(lcCse()) << raw.toBase64();
+    return ret;
+}
+
+std::string FolderMetadata::genMetadataPass() const {
+    const char* charmap = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+    const size_t charmapLength = strlen(charmap);
+    const int bytes = 16;
+    std::string result;
+    result.reserve(bytes);
+    generate_n(back_inserter(result), bytes, [&](){
+      return charmap[rand() % charmapLength];
```

Comment:
Especially `rand()` is not really cryptographic secure, as @rullzer pointed out :)